### PR TITLE
limit gc speed

### PR DESCRIPTION
**Pull request by july2993**
### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

File: pump/storage/storage.go [hunk 1]
```diff
@@ -519,34 +520,67 @@ func (a *Append) GCTS(ts int64) {
 	atomic.StoreInt64(&a.gcTS, ts)
 	a.saveGCTSToDB(ts)
 	gcTSGauge.Set(float64(oracle.ExtractPhysical(uint64(ts))))
+
+	if atomic.LoadInt32(&a.gcWorking) == 1 {
```
Comment (GregoryIan) [inline pump/storage/storage.go]:
```
if !atomic.CompareAndSwap(&a.gcWorking, 0, 1) {
    return
}

go func() {
    defer atomic.StoreInt32(&a.gcWorking, 0)
    a.doGCTS(ts - int64(oracle.EncodeTSO(maxTxnTimeoutSecond*1000)))
}
```

File: pump/storage/storage.go [hunk 2]
```diff
@@ -519,34 +520,67 @@ func (a *Append) GCTS(ts int64) {
...
+		}
+
+		irange := &util.Range{
+			Start: encodeTSKey(0),
+			Limit: encodeTSKey(ts + 1),
+		}
+		iter := a.metadata.NewIterator(irange, nil)
+
+		deleteBatch := 0
+		for iter.Next() && deleteBatch < 100 {
```
Comment (GregoryIan) [inline pump/storage/storage.go]: replace deleteBatch with batch.Len()

File: pump/storage/storage.go [hunk 3]
```diff
@@ -519,34 +520,67 @@ func (a *Append) GCTS(ts int64) {
...
+				deleteBatch++
+			}
+		}
+
+		if iter.Next() == false {
+			if batch.Len() > 0 {
+				err := a.metadata.Write(batch, nil)
+				if err != nil {
+					log.Error(err)
+				}
```
Comment (GregoryIan) [inline pump/storage/storage.go]: batch.Reset()

File: pump/storage/storage.go [hunk 4]
```diff
@@ -519,34 +520,67 @@ func (a *Append) GCTS(ts int64) {
...
-		err := a.metadata.Write(batch, nil)
+	for {
+		var stats leveldb.DBStats
+		err := a.metadata.Stats(&stats)
 		if err != nil {
 			log.Error(err)
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		if stats.LevelTablesCounts[0] >= l0Trigger {
```
Comment (GregoryIan) [inline pump/storage/storage.go]: output log

**Comment by GregoryIan** [comment]
Rest LGTM

File: pump/storage/storage.go (hunk 4 above)
Comment (july2993) [reply pump/storage/storage.go]: add in https://github.com/pingcap/tidb-binlog/pull/620/commits/a82e66fc7c7ec709168ea5211a82cc9cf0de5217

**Pull request merged**